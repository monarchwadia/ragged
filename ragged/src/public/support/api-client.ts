export * from "../../../src/support/ApiClient.js";