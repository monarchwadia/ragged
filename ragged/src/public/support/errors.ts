export * from "../../../src/support/CustomErrors.js";