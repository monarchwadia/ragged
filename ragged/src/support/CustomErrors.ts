import { Logger } from "./logger/Logger.js";

/**
 * Custom errors for drivers. Drivers can throw these errors to indicate
 * that something went wrong.
 */
export class BaseCustomError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "BaseError";
    }
}

/**
 * Thrown when a driver fails to parse JSON.
 */
export class JsonParseError extends BaseCustomError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "JsonParseError";
    }
}

export class JsonStringifyError extends BaseCustomError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "JsonStringifyError";
    }
}

/**
 * Throw when a request fails with a non-2xx status code.
 */
export class FetchRequestFailedError extends BaseCustomError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "FetchRequestFailedError";
    }
}

export class FetchResponseNotOkError extends BaseCustomError {
    private static logger: Logger = new Logger("FetchResponseNotOkError");
    constructor(public response: Response, public status: number) {
        const errorText = `Received a non-200 response from an API call while doing API call to ${response.url}. Status was ` + status + "."
        super(errorText);
        this.name = "FetchResponseNotOkError";
        response.text()
            .then((txt) => FetchResponseNotOkError.logger.info(errorText + "\r\n" + txt))
            .catch((e) => {
                FetchResponseNotOkError.logger.error(`${errorText}\r\n`, e);
            });
    }
}

export class MappingError extends BaseCustomError {
    constructor(message: string, public cause?: unknown) {
        super(message);
        this.name = "MappingError";
    }
}

export class UnknownError extends BaseCustomError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "UnknownError";
    }
}

export class ParameterValidationError extends BaseCustomError {
    constructor(message: string) {
        super(message);
        this.name = "ParameterError";
    }
}

export class NotImplementedError extends BaseCustomError {
    constructor(message: string) {
        super(message);
        this.name = "NotImplementedError";
    }
}

export class RetryError extends BaseCustomError {
    constructor(message: string) {
        super(message);
        this.name = "RetryError";
    }
}
