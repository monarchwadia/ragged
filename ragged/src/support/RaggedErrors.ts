import { Logger } from "./logger/Logger";

/**
 * Errors for drivers and more.
 */
export class BaseRaggedError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "BaseError";
    }
}

/**
 * Thrown when a driver fails to parse JSON.
 */
export class JsonParseError extends BaseRaggedError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "JsonParseError";
    }
}

export class JsonStringifyError extends BaseRaggedError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "JsonStringifyError";
    }
}

/**
 * Throw when a request fails with a non-2xx status code.
 */
export class FetchRequestFailedError extends BaseRaggedError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "FetchRequestFailedError";
    }
}

export class FetchResponseNotOkError extends BaseRaggedError {
    private static logger: Logger = new Logger("FetchResponseNotOkError");
    constructor(public response: Response, public status: number) {
        const errorText = `Received a non-200 response from an API call while doing API call to ${response.url}. Status was ` + status + "."
        super(errorText);
        this.name = "FetchResponseNotOkError";
        response.text()
            .then((txt) => FetchResponseNotOkError.logger.info(errorText + "\r\n" + txt))
            .catch((e) => {
                FetchResponseNotOkError.logger.error(`${errorText}\r\n`, e);
            });
    }
}

export class MappingError extends BaseRaggedError {
    constructor(message: string, public cause?: unknown) {
        super(message);
        this.name = "MappingError";
    }
}

export class UnknownError extends BaseRaggedError {
    constructor(message: string, public cause: unknown) {
        super(message);
        this.name = "UnknownError";
    }
}

export class ParameterValidationError extends BaseRaggedError {
    constructor(message: string) {
        super(message);
        this.name = "ParameterError";
    }
}

export class NotImplementedError extends BaseRaggedError {
    constructor(message: string) {
        super(message);
        this.name = "NotImplementedError";
    }
}

export class RetryError extends BaseRaggedError {
    constructor(message: string) {
        super(message);
        this.name = "RetryError";
    }
}

export class InstantiationError extends BaseRaggedError {
    constructor(message: string) {
        super(message);
        this.name = "InstantiationError";
    }
}