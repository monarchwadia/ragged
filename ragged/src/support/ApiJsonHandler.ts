import { JsonParseError, JsonStringifyError } from "./RaggedErrors";
import { Logger } from "./logger/Logger";

export class ApiJsonHandler {
    static logger: Logger = new Logger('AdapterJsonHandler');

    static parse(json: string): any {
        ApiJsonHandler.logger.debug(`Parsing JSON: ${json}`);
        try {
            return JSON.parse(json);
        } catch (e) {
            const err = new JsonParseError("Failed to parse JSON.", e);
            throw err;
        }
    }

    static stringify(obj: any): string {
        ApiJsonHandler.logger.debug(`Stringifying object to JSON:`, obj);
        try {
            return JSON.stringify(obj);
        } catch (e) {
            const err = new JsonStringifyError("Failed to stringify JSON.", e);
            throw err;
        }
    }

    static async parseResponse(response: Response): Promise<any> {
        ApiJsonHandler.logger.debug(`Parsing response:`, response);
        try {
            const text = await response.text();
            return ApiJsonHandler.parse(text);
        } catch (e) {
            const err = new JsonParseError("Failed to parse JSON response.", e);
            throw err;
        }
    }
}
