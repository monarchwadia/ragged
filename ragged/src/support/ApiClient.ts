import { ApiJsonHandler } from "./ApiJsonHandler.js";
import { FetchRequestFailedError, FetchResponseNotOkError } from "./CustomErrors.js";
import { Logger } from "./logger/Logger.js";


type PostOpts = {
    headers?: RequestInit["headers"];
    body?: any;
}

/**
 * This class is passed into each driver to allow the driver to make API calls.
 * This is useful for testing, as we can mock this class to return a fake response.
 * Every driver will use this class to make RESTful API calls.
 */
export class ApiClient {
    static logger: Logger = new Logger('ApiClient');

    async get(url: string, request: PostOpts): Promise<any> {
        return this.doCall("GET", url, request);
    }

    async post(url: string, request: PostOpts): Promise<any> {
        return this.doCall("POST", url, request);
    }

    private async doCall(method: string, url: string, request: PostOpts): Promise<Response> {
        const requestInit: RequestInit = {}

        requestInit.method = method;

        if (request.headers) {
            requestInit.headers = request.headers;
        }

        if (request.body) {
            requestInit.body = ApiJsonHandler.stringify(request.body)
        }

        let response: Response | null = null;
        try {
            response = await fetch(url, requestInit);
        } catch (e) {
            throw new FetchRequestFailedError("Failed to make fetch call.", e);
        }

        if (!response.ok) {
            throw new FetchResponseNotOkError(response, response.status);
        }

        return ApiJsonHandler.parseResponse(response);
    }
}