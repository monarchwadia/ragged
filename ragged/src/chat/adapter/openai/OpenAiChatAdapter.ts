import { ApiClient } from "../../../support/ApiClient";
import { BaseChatAdapter, ChatAdapterRequest } from "../BaseChatAdapter.types";
import { OpenAiChatCompletionRequestBody, OpenAiChatCompletionResponseBody } from "./OpenAiApiTypes";
import { mapFromOpenAi, mapToOpenAi } from "./OpenAiChatMappers";


export type OpenAiChatAdapterConfig = {
    apiKey?: string | undefined;
    organizationId?: string | undefined;
    rootUrl?: string;
}

export class OpenAiChatAdapter implements BaseChatAdapter {
    private apiKey: string | undefined;
    private organizationId: string | undefined;
    private rootUrl: string = "https://api.openai.com/v1/chat/completions";

    constructor(private driverApiClient: ApiClient, config: OpenAiChatAdapterConfig = {}) {
        if (config.apiKey) {
            this.apiKey = config.apiKey;
        }

        if (config.organizationId) {
            this.organizationId = config.organizationId;
        }

        if (config.rootUrl) {
            this.rootUrl = config.rootUrl;
        }
    }

    async chat(request: ChatAdapterRequest) {
        const mappedRequest = mapToOpenAi(request);
        const apiResponse = await this.chatCompletion(mappedRequest);
        const mappedResponse = mapFromOpenAi(apiResponse.json);
        return {
            history: mappedResponse,
            raw: apiResponse.raw
        
        };
    }


    private async chatCompletion(body: OpenAiChatCompletionRequestBody) {
        const headers: HeadersInit = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.apiKey}`
        };

        if (this.organizationId) {
            headers['OpenAI-Organization'] = this.organizationId;
        }

        return await this.driverApiClient.post(
            this.rootUrl,
            {
                headers,
                body
            }
        )
    }
}