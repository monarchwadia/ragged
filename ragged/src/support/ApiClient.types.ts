import { ApiClient } from "./ApiClient";

export type ApiClientFactory = () => ApiClient;