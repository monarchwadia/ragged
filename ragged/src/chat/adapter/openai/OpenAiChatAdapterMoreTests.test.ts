import { startPollyRecording } from "../../../test/startPollyRecording";
import { ApiClient } from "../../../support/ApiClient";
import { OpenAiChatAdapter as OpenAiChatDriver } from "./OpenAiChatAdapter";
import { objToReadableStream } from "../../../test/objectToReadableStream";

// TODO: merge with OpenAiChatAdapter.test.ts
describe("OpenAiChatDriver", () => {
  let apiClient: ApiClient;
  let driver: OpenAiChatDriver;

  beforeEach(() => {
    const config = {
      // apiKey: process.env.OPENAI_API_KEY,
    };

    apiClient = new ApiClient();
    driver = new OpenAiChatDriver(apiClient, config);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should pass the body directly into the apiclient", () => {
    const spy = jest.spyOn(apiClient, "post").mockImplementation(() =>
      Promise.resolve({
        json: {},
        raw: {
          request: new Request("https://not-real.com"),
          response: new Response(new ReadableStream()),
        },
      })
    );

    driver.chat({ model: "gpt-3.5-turbo", history: [] });

    expect(spy).toHaveBeenCalledWith(
      "https://api.openai.com/v1/chat/completions",
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer undefined`,
        },
        body: { model: "gpt-3.5-turbo", messages: [], tools: undefined },
      }
    );
  });

  it("should perform a chat completion", async () => {
    const polly = startPollyRecording(
      "OpenAiChatDriver > should perform a chat completion"
    );
    const result = await driver.chat({
      model: "gpt-3.5-turbo",
      history: [
        { type: "system", text: "You are a chatbot." },
        { type: "user", text: "What are you?" },
      ],
    });
    await polly.stop();

    expect(result.history).toMatchInlineSnapshot(`
      [
        {
          "text": "I am a chatbot, a computer program designed to simulate conversation with human users. How can I assist you today?",
          "type": "bot",
        },
      ]
    `);
  });
});
