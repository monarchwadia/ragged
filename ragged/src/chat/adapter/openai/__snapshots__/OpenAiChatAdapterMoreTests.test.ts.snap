// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAiChatDriver should perform a chat completion 1`] = `
{
  "history": [
    {
      "text": "I am a chatbot, a computer program designed to simulate conversation with human users. How can I assist you today?",
      "type": "bot",
    },
  ],
}
`;
